NewsPanel: Add news as a builtin panel

ryantxu opened the pull request:
This PR adds news as a builtin (currently alpha) panel.

It has a semi anti-pattern that the panel loads data directly... but I think it is an OK first step until we create a more robust RSS datasource and have a builtin RSS feed

ryantxu committed:
add news panel

ryantxu committed:
new panel

ryantxu committed:
update yarn.lock

briancurt committed:
Cloudwatch ECS Container Insights Support (#21125)

dprokop commented:
@ryantxu is this a replacement for the https://github.com/grafana/rss-panel? Not super wild about the `NewsPanel` naming. Think Rss Panel would work better.

hugohaggmark committed:
e2e: Updates truth image (#21132)

hugohaggmark commented:
> @ryantxu is this a replacement for the https://github.com/grafana/rss-panel? Not super wild about the `NewsPanel` naming. Think Rss Panel would work better.

I agree with Dominik about the naming and also agree that the async data loading should be moved sometime in the future using Redux state and dispatching actions.

davkal committed:
Prometheus: Display HELP and TYPE of metrics if available (#21124)

* Prometheus: Display HELP and TYPE of metrics if available

- Prometheus recently added a metadata API around HELP and TYPE of
metrics
- request metadata when datasource instance is created
- use metadata to show help and type in typeahead suggestions and in
metrics selector as tooltip

* Fix types

ivanahuckova committed:
Fix log row when query is short (#21126)

boyskila committed:
Chore: Upgrade react, react-dom, react-test-renderer versions (#21130)

* Chore: Upgrade react, react-dom, react-test-renderer versions

* Chore: Normalize windows path for webpack plugin test

* Chore: Replace deprecated react-testing-hooks-library

torkelo commented:
The reason I like New Panel as the name for this panel is that at its core (and future) this will be a pure visualization panel, that shows data as news items (with time, heading, author, body). So it won't really handle any of the rss in the future as I see this being handled by a built-in RSS data source. 

Having the rss loading being done by the panel it'self is a temporary solution so that we can actually deliver this panel (something we have promised to marketing for some time)

hugohaggmark commented:
> The reason I like New Panel as the name for this panel is that at its core (and future) this will be a pure visualization panel, that shows data as news items (with time, heading, author, body). So it won't really handle any of the rss in the future as I see this being handled by a built-in RSS data source.
> 
> Having the rss loading being done by the panel it'self is a temporary solution so that we can actually deliver this panel (something we have promised to marketing for some time)

That piece of information was great. I take back my previous comment completely.

dprokop commented:
Thanks Torkel for clarification

torkelo committed:
NewsPanel: initial design

torkelo commented:
Initial design: 

<img width="1608" alt="Screen Shot 2019-12-17 at 21 59 29" src="https://user-images.githubusercontent.com/10999/71034167-1812b680-2119-11ea-92d5-3f2e3f41fda7.png">

ryantxu commented:
Re news vs RSS -- similar to logs, I hope any datasource that returns title,link,content,etc could be a "news" source, not just RSS... BUT I don't want to hold this panel up longer waiting for:
1. a good general RSS data source
2. a way to deliver Grafana with out-of-the-box builtin datasources (an RSS datasource pointing to grafana blog content)

hugohaggmark committed:
e2e: Migrates query variable CRUD tests to new framework (#21146)

* Refactor: Adds params to visit

* Refactor: Restructures exported Pages somewhat

* Refactor: Moves more into new framework but holdup because of bugs in digest

* Refactor: Finish migrating templating e2e tests to new framework

* Refactor: Changes after merge with master

* Refactor: Removes weird change

* Refactor: Adds duplication test

* Refactor: Adds move down and move up variable asserts

* Refactor: Adds some test to value select dropdown

peterholmberg committed:
grafana/ui: New table component (#20991)

* first working example

* Support sorting, adding types while waiting for official ones

* using react-window for windowing

* styles via emotion

* sizing

* set an offset for the table

* change table export

* fixing table cell widths

* Explore: Use new table component in explore (#21031)

* Explore: Use new table component in explore

* enable oncellclick

* only let filterable columns be clickable, refactor renderrow

* remove explore table

* Keep using old merge tables logic

* prettier

* remove unused typings file

* fixing tests

* Fixed explore table issue

* NewTable: Updated styles

* Fixed unit test

* Updated TableModel

* Minor update to explore table height

* typing

tehGoti committed:
PagerDuty: Fix custom_details to be a JSON object instead of a string (#21150)

torkelo committed:
Dashboard: Only show resize-handle on hover (#21160)

dprokop committed:
DataLinks: Sanitize data/panel link URLs (#21140)

* Sanitize html in panel links

* Add sanitize-url package

* Enable config mocking

* Sanitize datalinks urls

* Update public/app/core/config.ts

* Minor test update

* Remove sanitize-url dependency

* Remove typings

* Review update

* Revert "Remove sanitize-url dependency"

This reverts commit c4f38e6de64b36fb398f0dc61906ae070d69c5ed.

* Revert "Remove typings"

This reverts commit 676d47e8c2e46d6eeb57003d0a7e0fd99e75310f.

* Sanitaze, don't escape html when sanitizing URL

hugohaggmark committed:
Table: Matches column names with unescaped regex characters (#21164)

* Table: Matches column names with unescaped regex characters
Fixes #21106

* Chore: Cleans up unused code

xlson committed:
docs: always updates docker image before building docs site (#21165)

davkal committed:
Prometheus: Fix typehead after binary operators (#21152)

- no longer do right-trim to prevent suggestions after space
- rely on splitting on syntax elements to identify relevant prefix
- added tests

peterholmberg committed:
Panel: Show inspect panel in Drawer instead of Modal (#21148)

michael-az committed:
Alerting: Add more information to webhook notifications (#20420)

Adds the orgId, dashboardId and panelId of the Webhook alert notifier.

torkelo committed:
Style update

torkelo committed:
Update

torkelo committed:
DashboardGrid: Change grid margin to 8, to align to 8px grid (#21167)

aknuds1 committed:
Changelog: Add PagerDuty breaking change (#21170)

hugohaggmark committed:
Docs: Adds best practices after visit and a link back to e2e.md (#21117)

* Docs: Adds best practices after visit and a link back to e2e.md in grafana-e2e

* Update contribute/style-guides/e2e.md

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

* Update packages/grafana-e2e/README.md

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

* Docs: Changes after PR comments

* Update contribute/style-guides/e2e.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

* Update contribute/style-guides/e2e.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

* Docs: Changes after PR comments

sebimarkgraf committed:
grafana/toolkit: Add option to override webpack config (#20872)

* Toolkit: Add possibility to add custom webpack config

* Toolkit: Refactor webpack to utilize async-await

* Toolkit: Rename config file and allow named export

yesoreyeram committed:
[docs] Azure monitor link in templating (#21173)

Docs: Added missing azure monitor reference in Templating topic.

stevenvachon committed:
@grafana/toolkit: webpack extend TS→JS (#21176)

ryantxu committed:
remove async comment

ryantxu committed:
merge master

dprokop commented on public/app/plugins/panel/news/NewsPanel.tsx:
```diff
@@ -0,0 +1,128 @@
+  isError?: boolean;
+}
+
+const CORS_PROXY = 'https://cors-anywhere.herokuapp.com/';
```
Ha, never heard of this, amazing!

ryantxu replied:
I'm a bit nervous about adding it automatically though -- after we have lenient CORS headers set on our feed, I think we should just recommend it in the UI rather than adding it automatically

dprokop replied:
Yeah, that's a good thinking

ryantxu committed:
Merge remote-tracking branch 'origin/master' into news-panel

ryantxu committed:
fix lint

ryantxu committed:
don't update docs

ryantxu commented:
Anyone have ideas on why the tests always break on:
![image](https://user-images.githubusercontent.com/705951/71154101-9e013100-21ef-11ea-93d0-8be82d5305b1.png)

```
throw new Error("StringMap expected string key");
```
perhaps the `rss-feed` changes some imports?

torkelo commented:
very strange, something in NewPanel.tsx is causing the ng-annotate plugin to crash, trying remove parts piece by piece to see what is causing it

ryantxu committed:
Merge remote-tracking branch 'origin/master' into news-panel

ryantxu committed:
add proxy flag

torkelo commented:
It's the rss-parser that is causing the test failure, think we will have to load it async or find another library. Sadly I had a lot of trouble with other RSS feeds (http://feeds.arstechnica.com/arstechnica/index/), many threw parsing errors. Not sure why, no open related issues on the rss parser repo.

ryantxu committed:
implement rss without library

ryantxu committed:
fewer fields

ryantxu committed:
no generated

ryantxu commented:
OK -- I got rid of the RSS library and am just parsing it directly with DOMParser:
https://github.com/grafana/grafana/pull/21128/files#diff-89707cb3779780ddefd92c6758248becR16

likely flaky and will only work with feeds shaped exactly like this... we can/should replace with a more robust server-side datasource

torkelo commented:
I have a new home dashboard that looks like this:
![image](https://user-images.githubusercontent.com/10999/71256279-867d9180-2330-11ea-882a-a4d7fd1039b5.png)

Should we merge this as is and work on new default home dashboard in seperate PR? 

Todo for that is 
* Add utm_source attribute to urls for tracking
* Would be good with a custom grafana.com rss url to use for this so we can have more control over what is shown there.

mckn commented on packages/grafana-ui/src/themes/mixins.ts:
```diff
@@ -0,0 +1,21 @@
+import { GrafanaTheme } from '@grafana/data';
+
+export function cardChrome(theme: GrafanaTheme): string {
```
Are we planning on using this styling more places then in the `NewsPanel`? If not I would probably move it to a file located close the the component.

torkelo replied:
yes, this we use card styles in many places (in sass), but this the emotion equivalent to a sass mixing :)

mckn replied:
Super, got that! But then it feels like it is placed in the correct place since it will be globally used. Looks good! :)

dprokop replied:
@mckn we want to intoduce some reusable pieces of styles as we see more needs for that as there is more emotion code in the codebase

mckn commented on public/app/plugins/panel/news/NewsPanelEditor.tsx:
```diff
@@ -0,0 +1,57 @@
+      <>
+        <PanelOptionsGroup title="Feed">
+          <div className="gf-form">
+            <FormField
```
One super small improvement could be to use the new `Forms` components located in `@grafana-ui/Forms/*`

dprokop replied:
@mckn not yet unfortunately. This is Panel edit and I think we want to keep it consistent with other panel edits.

mckn replied:
OK ok, sweet! :)

mckn commented on public/app/plugins/panel/news/rss.ts:
```diff
@@ -0,0 +1,28 @@
+import { RssFeed, RssItem } from './types';
+
+const CORS_PROXY = 'https://cors-anywhere.herokuapp.com/';
```
Is this a trusted proxy? Feels a bit dangerous to use this one otherwise. If you add an feedUrl containing sensitive information this proxy will have access to all that information.

ryantxu replied:
Good point @mckn -- my plan was to remove this when our feed has the right CORS headers set and *then* move this panel from 'alpha' to general release.

But, I went ahead and changed the structure so we explicitly show the "herokuapp.com" in the default URL and will replace it when our server is setup for this.

mckn reviewed (commented):
It looks really good to me. I have added some super small thoughts/questions. The major thing is the question regarding the proxy. Apart from that oen it is a go from me. Well done! 🙌

ryantxu committed:
Merge remote-tracking branch 'origin/master' into news-panel

ryantxu committed:
make proxy explicit

ryantxu committed:
make proxy explicit

ryantxu committed:
Merge remote-tracking branch 'origin/master' into news-panel

ryantxu committed:
update test snapshot
